## Data Mining UniProj - no.1 
#### Sobhan Moradiyan Daghigh & Fatemeh Chaji
##### 11/25/2020

In [1]:
# !pip install import_ipynb

In [2]:
import pandas as pd
import numpy as np
import import_ipynb
from FarsiToEngConverte import convert_en_characters

importing Jupyter notebook from FarsiToEngConverte.ipynb


### Reading the databases

In [3]:
product = pd.read_excel(r"./dataset/product.xlsx")
order_date = pd.read_csv(r"./dataset/tarikhche kharid.csv")
comments = pd.read_excel(r"./dataset/comment.xlsx")
orders = pd.read_csv(r"./dataset/orders.csv")
quality = pd.read_excel(r"./dataset/keifiat.xlsx")

### Define som functions in order to handle missed values 

#### We wrote "FarsiToEngConverte" file which convert Farsi chars to English

 ##### Look how beautiful it is ;)

In [4]:
convert_en_characters("گوشی موبایل سامسونگ J7Pro")

'goshi mobail samsong J7Pro'

In [5]:
def fill_in_missed_values(table, ref_col_no, goal_col_no):
    product_manipulated = pd.DataFrame(table) # In order to make a copy of it ..
    for i in range(table.shape[0]):
        if str(table.loc[i, goal_col_no]) == "nan":
            product_manipulated.loc[i, goal_col_no] = convert_en_characters(table.loc[i, ref_col_no])
    return product_manipulated

In [6]:
def reporter(table, table_manipulated, strategy):
        
    all_records    = table.shape[0]
    all_columns    = table.shape[1]
    
    # strategy 1 or 2
    if strategy == 12:
        difective_records_before_fill_in = sum([True for idx,row in table.iterrows() if any(row.isnull())])
        difective_records_after_fill_in  = sum([True for idx,row in table_manipulated.iterrows() if any(row.isnull())])
        print("For this database:   columns: %d   records: %d   difective_records_before_fill_in: %d   difective_records_after_fill_in: %d \n So the percentage of completeness is: %f"
              %(all_columns, all_records, difective_records_before_fill_in, difective_records_after_fill_in, np.divide(np.subtract(all_records, difective_records_after_fill_in), all_records)))

    # strategy 3
    elif strategy == 3:
        missed_values_before_fill_in = table.isnull().sum().sum()
        missed_values_after_fill_in  = table_manipulated.isnull().sum().sum()
        print("For this database:   columns: %d   records: %d   missed_values_before_fill_in: %d   missed_values_after_fill_in: %d \n So the percentage of completeness is: %f"
              %(all_columns, all_records, missed_values_before_fill_in, missed_values_after_fill_in, np.subtract(1, np.divide(missed_values_after_fill_in, np.multiply(all_columns, all_records)))))     

### Ok now, we wanna fill in the missing values

#### 1- Product database

In [7]:
product.head()

id                                   product_title_fa  \
0  723302            ماگ حرارتی لومانا مدل عرفانه کد MAG1042   
1  727089       ست آبکش و آبگیر تیتیز مدل AP-9141 سایز متوسط   
2  727096  روغن مرطوب کننده کدلی مدل Divine oil حجم 50 می...   
3  727803  ست لحاف پوپلین هوبی مدل اما یکنفره چهار تکه کاوری   
4  725222                         ماگ آبنبات رنگی مدل AR0143   

                         product_title_en  \
0                                     NaN   
1                                     NaN   
2                                     NaN   
3  Hobby Emma Quilt Single Set - 4 Pieces   
4                                     NaN   

                                            url_code  \
0            ماگ-حرارتی-لومانا-مدل-عرفانه-کد-mag1042   
1       ست-آبکش-و-آبگیر-تیتیز-مدل-ap-9141-سایز-متوسط   
2  روغن-مرطوب-کننده-کدلی-مدل-divine-oil-حجم-50-می...   
3  ست-لحاف-پوپلین-هوبی-مدل-اما-یکنفره-چهار-تکه-کاوری   
4                         ماگ-آبنبات-رنگی-مدل-ar0143   

                                           title_alt   category_title_fa  \
0  mug، لیوان ، ماگ ، فنجان ، لیوان دسته دار ، لو...  ماگ، لیوان و فنجان   
1  ;hsi, steel, کاسه، کاسه بامبو، لگن، کالا خانه ...        آبکش و آبگیر   
2                                                NaN           روغن صورت   
3  روتختی اسپرت، اما، emma، روتختی یکنفره، لحاف ی...          سرویس خواب   
4      ماگ آبنبات رنگی،شذدشذشفقشدله،Mag Abnabatrangi  ماگ، لیوان و فنجان   

              category_keywords brand_name_fa  brand_name_en  \
0  ماگ، لیوان و فنجان CupandMug        لومانا         Lomana   
1       آبکش و آبگیر VirgoBasin         تیتیز          Titiz   
2            روغن صورت Face-Oil          کدلی       CAUDALIE   
3           سرویس خواب SleepSet          هوبی          Hobby   
4  ماگ، لیوان و فنجان CupandMug        متفرقه  Miscellaneous   

                                  product_attributes  
0  [{"Key":"وزن بسته‌بندی","Value":"480"},{"Key":...  
1  [{"Key":"دستگیره","Value":"1 جفت"},{"Key":"نوع...  
2  [{"Key":"حجم","Value":"50"},{"Key":"نوع عصاره"...  
3  [{"Key":"ابعاد رو بالشی","Value":"50×70"},{"Ke...  
4  [{"Key":"در","Value":"ندارد"},{"Key":"جنس","Va...

In [8]:
product.isnull().sum()

id                        0
product_title_fa          0
product_title_en      78835
url_code                 17
title_alt             22568
category_title_fa         0
category_keywords       260
brand_name_fa             0
brand_name_en             0
product_attributes    96950
dtype: int64

#### We fill in the missing datas in th "product_title_en" column for Product database with accepting a low percentage of errors while converting the characers

In [9]:
product_manipulated = fill_in_missed_values(product, "product_title_fa", "product_title_en")
product_manipulated.head()

id                                   product_title_fa  \
0  723302            ماگ حرارتی لومانا مدل عرفانه کد MAG1042   
1  727089       ست آبکش و آبگیر تیتیز مدل AP-9141 سایز متوسط   
2  727096  روغن مرطوب کننده کدلی مدل Divine oil حجم 50 می...   
3  727803  ست لحاف پوپلین هوبی مدل اما یکنفره چهار تکه کاوری   
4  725222                         ماگ آبنبات رنگی مدل AR0143   

                                    product_title_en  \
0            mag hrarti lomana mdl arfanh kd MAG1042   
1      st abksh o abgir titiz mdl AP-9141 saiz mtost   
2  roghn mrtob knndh kdli mdl Divine oil hjm 50 m...   
3             Hobby Emma Quilt Single Set - 4 Pieces   
4                         mag abnbat rngi mdl AR0143   

                                            url_code  \
0            ماگ-حرارتی-لومانا-مدل-عرفانه-کد-mag1042   
1       ست-آبکش-و-آبگیر-تیتیز-مدل-ap-9141-سایز-متوسط   
2  روغن-مرطوب-کننده-کدلی-مدل-divine-oil-حجم-50-می...   
3  ست-لحاف-پوپلین-هوبی-مدل-اما-یکنفره-چهار-تکه-کاوری   
4                         ماگ-آبنبات-رنگی-مدل-ar0143   

                                           title_alt   category_title_fa  \
0  mug، لیوان ، ماگ ، فنجان ، لیوان دسته دار ، لو...  ماگ، لیوان و فنجان   
1  ;hsi, steel, کاسه، کاسه بامبو، لگن، کالا خانه ...        آبکش و آبگیر   
2                                                NaN           روغن صورت   
3  روتختی اسپرت، اما، emma، روتختی یکنفره، لحاف ی...          سرویس خواب   
4      ماگ آبنبات رنگی،شذدشذشفقشدله،Mag Abnabatrangi  ماگ، لیوان و فنجان   

              category_keywords brand_name_fa  brand_name_en  \
0  ماگ، لیوان و فنجان CupandMug        لومانا         Lomana   
1       آبکش و آبگیر VirgoBasin         تیتیز          Titiz   
2            روغن صورت Face-Oil          کدلی       CAUDALIE   
3           سرویس خواب SleepSet          هوبی          Hobby   
4  ماگ، لیوان و فنجان CupandMug        متفرقه  Miscellaneous   

                                  product_attributes  
0  [{"Key":"وزن بسته‌بندی","Value":"480"},{"Key":...  
1  [{"Key":"دستگیره","Value":"1 جفت"},{"Key":"نوع...  
2  [{"Key":"حجم","Value":"50"},{"Key":"نوع عصاره"...  
3  [{"Key":"ابعاد رو بالشی","Value":"50×70"},{"Ke...  
4  [{"Key":"در","Value":"ندارد"},{"Key":"جنس","Va...

In [10]:
product_manipulated.isnull().sum()

id                        0
product_title_fa          0
product_title_en          0
url_code                 17
title_alt             22568
category_title_fa         0
category_keywords       260
brand_name_fa             0
brand_name_en             0
product_attributes    96950
dtype: int64

#### Congratulation :)

#### But unfortunately theres a large amount of missing data in column "product_attributes" which we cant do sth for them ..
#### In addition its a necessary attr for out database, so we can't ignore the attr

#### So we decide to measure the amount of completeness using 3rd strategy: 
#####  A: attribute fields containing values | B: records × attributes

In [11]:
reporter(product, product_manipulated, 3)

For this database:   columns: 10   records: 100000   missed_values_before_fill_in: 119795   missed_values_after_fill_in: 119795 
 So the percentage of completeness is: 0.880205


#### just for getting the result of 1st strategy

In [12]:
reporter(product, product_manipulated, 12)

For this database:   columns: 10   records: 100000   difective_records_before_fill_in: 97736   difective_records_after_fill_in: 97736 
 So the percentage of completeness is: 0.022640


#### 2- Order_date database

In [13]:
order_date.head()

id  product_variant_id  selling_price  rrp_price base_price  \
0  18752043             1977385         850000     850000         \N   
1  18752022             1977380       12550000   12550000         \N   
2  18752046             1977380       12500000   12500000         \N   
3  18751982             1977374         310000     310000         \N   
4  18752018             1977374         300000     300000         \N   

  buy_price  order_limit start_at end_at tags  show_in_price_history  active  \
0        \N            1       \N     \N   \N                      1       1   
1        \N            2       \N     \N   \N                      1       0   
2        \N            2       \N     \N   \N                      1       1   
3        \N            2       \N     \N   \N                      1       0   
4        \N            2       \N     \N   \N                      1       1   

            created_at  product_id  marketplace_seller_id  
0  2018-11-26 18:59:52      667969                   2962  
1  2018-11-26 18:59:31      150144                  78525  
2  2018-11-26 18:59:55      150144                  78525  
3  2018-11-26 18:58:48      908541                   4199  
4  2018-11-26 18:59:28      908541                   4199

In [14]:
order_date_manipulated1 = order_date.replace(to_replace='\\N', value= np.nan)
order_date_manipulated1.head()

id  product_variant_id  selling_price  rrp_price base_price  \
0  18752043             1977385         850000     850000        NaN   
1  18752022             1977380       12550000   12550000        NaN   
2  18752046             1977380       12500000   12500000        NaN   
3  18751982             1977374         310000     310000        NaN   
4  18752018             1977374         300000     300000        NaN   

  buy_price  order_limit start_at end_at tags  show_in_price_history  active  \
0       NaN            1      NaN    NaN  NaN                      1       1   
1       NaN            2      NaN    NaN  NaN                      1       0   
2       NaN            2      NaN    NaN  NaN                      1       1   
3       NaN            2      NaN    NaN  NaN                      1       0   
4       NaN            2      NaN    NaN  NaN                      1       1   

            created_at  product_id  marketplace_seller_id  
0  2018-11-26 18:59:52      667969                   2962  
1  2018-11-26 18:59:31      150144                  78525  
2  2018-11-26 18:59:55      150144                  78525  
3  2018-11-26 18:58:48      908541                   4199  
4  2018-11-26 18:59:28      908541                   4199

In [15]:
order_date_manipulated1.isnull().sum()

id                             0
product_variant_id             0
selling_price                  0
rrp_price                      0
base_price               2090066
buy_price                2177495
order_limit                    0
start_at                 2127157
end_at                   2127157
tags                     2126456
show_in_price_history          0
active                         0
created_at                     0
product_id                     0
marketplace_seller_id          0
dtype: int64

In [16]:
order_date.shape[0] # count of records

2184358

#### We found that 5 columns at this database are almost null. We think they're not important attrs.
#### So for this database we secide to ignore thiese attrs and follow the 2nd strategy in completeness:
#### A: number of data required for the particular context in the data file    |    B: number of data in the specified particular context of intended use

In [17]:
order_date_manipulated2 = order_date_manipulated1.drop(columns=["base_price", "buy_price", "start_at", "end_at", "tags"])
order_date_manipulated2.head()

id  product_variant_id  selling_price  rrp_price  order_limit  \
0  18752043             1977385         850000     850000            1   
1  18752022             1977380       12550000   12550000            2   
2  18752046             1977380       12500000   12500000            2   
3  18751982             1977374         310000     310000            2   
4  18752018             1977374         300000     300000            2   

   show_in_price_history  active           created_at  product_id  \
0                      1       1  2018-11-26 18:59:52      667969   
1                      1       0  2018-11-26 18:59:31      150144   
2                      1       1  2018-11-26 18:59:55      150144   
3                      1       0  2018-11-26 18:58:48      908541   
4                      1       1  2018-11-26 18:59:28      908541   

   marketplace_seller_id  
0                   2962  
1                  78525  
2                  78525  
3                   4199  
4                   4199

In [18]:
order_date_manipulated2.isnull().sum()

id                       0
product_variant_id       0
selling_price            0
rrp_price                0
order_limit              0
show_in_price_history    0
active                   0
created_at               0
product_id               0
marketplace_seller_id    0
dtype: int64

In [19]:
reporter(order_date_manipulated1, order_date_manipulated2, 12)

For this database:   columns: 15   records: 2184358   difective_records_before_fill_in: 2180384   difective_records_after_fill_in: 0 
 So the percentage of completeness is: 1.000000


#### 3- Comments database

In [20]:
comments.head()

product_id         confirmed_at  \
0       90213  2018-08-23 11:05:47   
1       56871  2018-08-23 10:21:53   
2       49738  2018-10-13 11:13:39   
3      134367  2018-10-17 11:27:57   
4      151402  2018-08-23 10:28:47   

                                             comment  
0  سلام، قبل اینکه نظرم رو بگم میخواستم به یک موض...  
1  بوی تند ولی خوشبو داره.ماندگاریش خوبه و هر چقد...  
2  متاسفانه عمر مفید این ماشین کم هست و بعد از دو...  
3  افا از وقتی من اینو واسه پسرم خریدم همه بچه ها...  
4  این تیغ محصول کشور آمریکا هست ( در عکس محصول م...

In [21]:
comments.isnull().sum()

product_id        0
confirmed_at      0
comment         105
dtype: int64

#### So the most important attr in thid database is attr "comment" tha have 105 nulls. We cant ignore this important column so use the 1st strategy of completetness

In [22]:
reporter(comments, comments, 12)

For this database:   columns: 3   records: 100008   difective_records_before_fill_in: 105   difective_records_after_fill_in: 105 
 So the percentage of completeness is: 0.998950


#### 4- Orders database

In [23]:
orders.head()

ID_Order  ID_Customer  ID_Item    DateTime_CartFinalize  \
0   2714054       469662    21386  2015-10-15 08:50:56.000   
1  11104039      3063877   248497  2018-02-11 00:29:26.000   
2   4228130      3184893    50144  2016-06-14 00:30:08.000   
3  22225624      6888562    70208  2018-09-03 14:37:19.000   
4   4068771      2533490    67627  2016-05-21 11:51:02.000   

   Amount_Gross_Order city_name_fa  Quantity_item  
0            597982.0   محمود آباد            1.0  
1            980000.0       خرمدره            1.0  
2            229358.0         قرچک            1.0  
3             16514.0           قم            1.0  
4            133028.0        تهران            1.0

In [24]:
orders.isnull().sum()

ID_Order                 0
ID_Customer              0
ID_Item                  0
DateTime_CartFinalize    0
Amount_Gross_Order       0
city_name_fa             0
Quantity_item            0
dtype: int64

#### Pashmam ;) Out of any missed data.

In [25]:
reporter(orders, orders, 3)

For this database:   columns: 7   records: 200000   missed_values_before_fill_in: 0   missed_values_after_fill_in: 0 
 So the percentage of completeness is: 1.000000


#### 5- Quality database

In [26]:
quality.head()

product_id                                      product_title title_en  \
0        3692                        ماوس بی‌سیم لاجیتک مدل M325       IT   
1       90213  شارژر همراه شیاومی مدل NDY-02-AN با ظرفیت 1000...       AC   
2       59473              یدک پولیشر میکروفایبر مهسان مدل 20119       HW   
3      120499   گوشی موبایل هوآوی آنر مدل 5X KIW-L21 دو سیم‌کارت       MO   
4       67200  شارژر همراه شیائومی مدل Mi ظرفیت 5000 میلی آمپ...       AC   

   user_id  likes  dislikes verification_status        recommend  \
0   989472      0         0            verified               \N   
1  3862150      4         1            verified      recommended   
2   626843      1         0            verified  not_recommended   
3   786887      6        11            verified          no_idea   
4   854531     19         4            verified          no_idea   

                                 title  \
0                                  NaN   
1                         واقعاً عالیه   
2             خیلی سخت حوله اش در میاد   
3                   گوشی مقرون به صرفه   
4  ابعاد، استحکام و نگهداری شارژ مناسب   

                                             comment  \
0                  واقعا عالیه. من که ازش خیلی راضیم   
1  سلام، قبل اینکه نظرم رو بگم میخواستم به یک موض...   
2  گیره های فلزی خیلی سخت تا میشوند و لذا حوله را...   
3  همه چیز در رابطه با ظاهر این گوشی بسیار خوب اس...   
4  اگر ظرفیتش براتون کافیه حتما بخرید.\r\nیه شارژ...   

                                          advantages  \
0                                                NaN   
1  ["عمر طولانی\r","افت بسیار کم میزان شارژ\r","ا...   
2                                                NaN   
3  ["صفحه نمایش پرنور و با کیفیت\r","کیفیت بالای ...   
4    ["ظریف و زیبا و باریک\r","بدنه محکم و با دوام"]   

                                       disadvantages  
0                                                NaN  
1                                          ["ندارد"]  
2                                                NaN  
3  ["کیفیت پایین اسپیکر\r","حاشیه خالی زیر صفحه ن...  
4                                    ["ظرفیت متوسط"]

In [27]:
quality.isnull().sum()

product_id                 0
product_title              0
title_en                   0
user_id                    0
likes                      0
dislikes                   0
verification_status        0
recommend                  0
title                   3270
comment                  117
advantages             57759
disadvantages          65879
dtype: int64

In [28]:
quality.shape[0]

100000

#### Here we think that ""comment" and "title" attrs are important for us but the attrs "advantages" and "disadvantages" are not necessary attrs to fill in (can be null).
#### So we ignore that 2 columns but keep the "comment" and  "title" missed values. Use 2nd strategy for completeness

In [29]:
quality_manipulated = quality.drop(columns=["advantages", "disadvantages"])
quality_manipulated.head()

product_id                                      product_title title_en  \
0        3692                        ماوس بی‌سیم لاجیتک مدل M325       IT   
1       90213  شارژر همراه شیاومی مدل NDY-02-AN با ظرفیت 1000...       AC   
2       59473              یدک پولیشر میکروفایبر مهسان مدل 20119       HW   
3      120499   گوشی موبایل هوآوی آنر مدل 5X KIW-L21 دو سیم‌کارت       MO   
4       67200  شارژر همراه شیائومی مدل Mi ظرفیت 5000 میلی آمپ...       AC   

   user_id  likes  dislikes verification_status        recommend  \
0   989472      0         0            verified               \N   
1  3862150      4         1            verified      recommended   
2   626843      1         0            verified  not_recommended   
3   786887      6        11            verified          no_idea   
4   854531     19         4            verified          no_idea   

                                 title  \
0                                  NaN   
1                         واقعاً عالیه   
2             خیلی سخت حوله اش در میاد   
3                   گوشی مقرون به صرفه   
4  ابعاد، استحکام و نگهداری شارژ مناسب   

                                             comment  
0                  واقعا عالیه. من که ازش خیلی راضیم  
1  سلام، قبل اینکه نظرم رو بگم میخواستم به یک موض...  
2  گیره های فلزی خیلی سخت تا میشوند و لذا حوله را...  
3  همه چیز در رابطه با ظاهر این گوشی بسیار خوب اس...  
4  اگر ظرفیتش براتون کافیه حتما بخرید.\r\nیه شارژ...

In [30]:
quality_manipulated.isnull().sum()

product_id                0
product_title             0
title_en                  0
user_id                   0
likes                     0
dislikes                  0
verification_status       0
recommend                 0
title                  3270
comment                 117
dtype: int64

In [31]:
reporter(quality, quality_manipulated, 12)

For this database:   columns: 12   records: 100000   difective_records_before_fill_in: 70750   difective_records_after_fill_in: 3310 
 So the percentage of completeness is: 0.966900


In [32]:
# Yohaaaa Finished :)